### PR TITLE
DAPI-1313: React revamp of Attribute Groups Grid

### PR DESCRIPTION
**Pull request by olivierpontier**
**Description (for Contributor and Core Developer)**

DAPI-1313_preview

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/shared/layouts/tables/TableBody.tsx [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import styled from 'styled-components';
+
+const TableBody = styled.tbody`
```
Comment (rouliane) [inline src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/shared/layouts/tables/TableBody.tsx:3]: what is the point of this file (premature optimization ?...) ? Do you plan to add some styles ? If no, i'd recommend to delete it

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/shared/layouts/tables/TableFoot.tsx [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import styled from 'styled-components';
+
+const TableFoot = styled.tfoot`
```
Comment (rouliane) [inline src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/shared/layouts/tables/TableFoot.tsx:3]: same remark here (empty styled component)

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/shared/layouts/tables/TableHead.tsx [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import styled from 'styled-components';
+
+const TableHead = styled.thead`
```
Comment (rouliane) [inline src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/shared/layouts/tables/TableHead.tsx:3]: same remark here (empty styled component)

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/config/requirejs.yml [hunk 4]
```diff
@@ -232,7 +232,7 @@ config:
           module: pim/controller/attribute-group/create
           aclResourceId: pim_enrich_attributegroup_create
         pim_enrich_attributegroup_index:
-          module: pim/controller/common/index
+          module: pim/controller/attribute-group/index #pim/controller/common/index
```
Comment (rouliane) [inline src/Akeneo/Platform/Bundle/UIBundle/Resources/config/requirejs.yml:235]: comment to remove ?

**Review by rouliane — APPROVED** [review]
Some blank lines are missing at the end of some files. (wrong IDE configuration ?)

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/shared/layouts/tables/TableBody.tsx (hunk 1 above)
Comment (olivierpontier) [reply src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/shared/layouts/tables/TableBody.tsx:3]: The goal is to create a suite of consistent  and named components for displaying tables (Table, TableBody, TableCell, ...).
The syntax of styled-components does not allow  `const TableBody = styled.tbody`. It is a function that need a parameter, a literal template or an object.

The following examples produce same code:

```
const TableBody = styled.tbody'
    background-color: red;
';
```
```
const TableBody = styled.tbody({
    backgroundColor: 'red';
});
```

Comment (olivierpontier) [reply src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/shared/layouts/tables/TableBody.tsx:3]: Further more the syntax of styled-components is more simple for a basic component like this.
The classic way should be:

```
const TableBody: FC = ({children}) => <tbody>{children}</tbody>;
```

Or simply use the HTML code, but I lost the naming.

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/pages/AttributeGroupsIndex.tsx [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+            await load();
+            if (mounted.current) {
+                setShowPlaceholder(false);
+            }
+        })();
+    }, []);
+
+    return (
+        <>
+            <PageHeader breadcrumb={breadcrumb} userButtons={userButtons} buttons={buttons} showPlaceholder={showPlaceholder}>
```
Comment (juliensnz) [inline src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/pages/AttributeGroupsIndex.tsx:33]: could you please make it a compound component?

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/workspaces/shared/src/components/breadcrumb/Breadcrumb.tsx [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+interface Props {
+    children: ReactElement<ItemProps> | Array<ReactElement<ItemProps>>;
+}
+
+const Breadcrumb = ({children}: Props) => {
+    const count = Children.count(children);
+
+    return (
+        <div className='AknBreadcrumb'>
```
Comment (juliensnz) [inline src/Akeneo/Platform/Bundle/UIBundle/Resources/workspaces/shared/src/components/breadcrumb/Breadcrumb.tsx:12]: could you please use a styled component?

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/workspaces/shared/src/components/breadcrumb/BreadcrumbItem.tsx [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+    onClick?: () => void;
+    isLast?: boolean;
+}
+
+const BreadcrumbItem = ({children: label, onClick, isLast}: PropsWithChildren<Props>) => {
+    const className = 'AknBreadcrumb-item' + (isLast ? ' AknBreadcrumb-item--final' : '');
+
+    if (onClick) {
+        return (
+            <span onClick={onClick} className={className + ' AknBreadcrumb-item--routable'}>
```
Comment (juliensnz) [inline src/Akeneo/Platform/Bundle/UIBundle/Resources/workspaces/shared/src/components/breadcrumb/BreadcrumbItem.tsx:13]: could you please use styled components. We try to not put any BEM in the shared folder

File: src/Akeneo/Platform/Bundle/UIBundle/Resources/workspaces/shared/src/components/PageHeader.tsx [hunk 8]
```diff
@@ -0,0 +1,84 @@
...
+  children: title,
+  breadcrumb,
+  buttons,
+  userButtons,
+  state,
+  imageSrc,
+  showPlaceholder
+}) => (
+  <Header>
+    <div className="AknTitleContainer-line">
```
Comment (juliensnz) [inline src/Akeneo/Platform/Bundle/UIBundle/Resources/workspaces/shared/src/components/PageHeader.tsx:44]: if you put elements in the shared folder, could you please use styled components

**Review by juliensnz — COMMENTED** [review]
What about testing? I don't think unit testing this is sufficient for this feature.

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/buttons/AttributeGroupsCreateButton.tsx [hunk 9]
```diff
@@ -0,0 +1,13 @@
+import React, {FC} from 'react';
+import {PimView} from '@akeneo-pim-community/legacy-bridge';
+
+const AttributeGroupsCreateButton: FC = () => {
+    return (
+       <PimView
+           className=''
```
[the remark this answers is not in the document]
Comment (olivierpontier) [reply src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/buttons/AttributeGroupsCreateButton.tsx]: The create button does not need a className but it's a required Props for the PimView component

[the remark this answers is not in the document]
Comment (juliensnz) [reply src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/buttons/AttributeGroupsCreateButton.tsx]: What do you think about making it optional on the PIMView?

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/pages/AttributeGroupsIndex.tsx (hunk 5 above)
Comment (olivierpontier) [reply src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/pages/AttributeGroupsIndex.tsx]: Refactoring wrote in the commit 7ae8ad4

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/apps/AttributeGroupsApp.tsx [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+import {AkeneoThemeProvider} from '@akeneo-pim-community/shared';
+import {AttributeGroupsIndex} from '../pages';
+import {AttributeGroupsDataGridProvider} from "../components";
+
+const AttributeGroupsApp: FC = () => {
+    return (
+        <DependenciesProvider>
+            <AkeneoThemeProvider>
+                <div>
+                    <AttributeGroupsDataGridProvider>
```
[the remark this answers is not in the document]
Comment (olivierpontier) [reply src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/apps/AttributeGroupsApp.tsx]: It is providing more the page state than the datagrid. I renamed it into AttributeGroupsIndexProvider

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/models/AttributeGroup.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+        id: number;
+    };
+};
+
+type AttributeGroupCollection = {
+    [group: string]: AttributeGroup;
+};
+
+
+const fromAttributeGroupsCollection = (collection: AttributeGroupCollection): AttributeGroup[] => {
```
[the remark this answers is not in the document]
Comment (olivierpontier) [reply src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/models/AttributeGroup.ts]: renamed as `toSortedAttributeGroupsArray` in the commit: 454076b

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/datagrids/attribute-groups/AttributeGroupsDataGrid.tsx [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+          (async () => saveOrder())();
+        }}
+      >
+        {groups.map((group) => (
+          <DataGrid.Row
+            key={group.code}
+            data={group}
+          >
+            <DataGrid.Column>
+              <ColumnLabel>{getLabel(group)}</ColumnLabel>
```
Comment (juliensnz) [inline src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/datagrids/attribute-groups/AttributeGroupsDataGrid.tsx:44]:
```suggestion
              <CellLabel>{getLabel(group)}</CellLabel>
```

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/datagrids/attribute-groups/AttributeGroupsDataGrid.tsx [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+        onRowMoveEnd={() => {
+          (async () => saveOrder())();
+        }}
+      >
+        {groups.map((group) => (
+          <DataGrid.Row
+            key={group.code}
+            data={group}
+          >
+            <DataGrid.Column>
```
Comment (juliensnz) [inline src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/datagrids/attribute-groups/AttributeGroupsDataGrid.tsx:43]:
```suggestion
            <DataGrid.Cell>
```

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/shared/datagrids/DataGrid.tsx [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+const DataGrid = <T extends {}>({
+  children,
+  isDraggable,
+  dataSource,
+  handleAfterMove,
+  compareData,
+}: PropsWithChildren<Props<T>>) => {
+  return (
+    <DataGridStateProvider
+      isDraggable={isDraggable || false}
```
Comment (juliensnz) [inline src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/shared/datagrids/DataGrid.tsx:25]: What do you think about putting the default value in the props spread instead?

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/shared/datagrids/HeaderRow.tsx [hunk 15]
```diff
@@ -0,0 +1,20 @@
+import React, {FC} from 'react';
+import {TableHead, TableHeadCell, TableHeadRow} from '../layouts/tables';
+
+type Props = {};
+
+const HeaderRow: FC<Props> = ({children}) => {
+  const isDraggable = true;
```
Comment (juliensnz) [inline src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/shared/datagrids/HeaderRow.tsx:7]: why that?

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/shared/datagrids/Row.tsx [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+      ref={rowRef}
+      isDragged={isDragged(data)}
+      onClick={() => {
+        if (handleClick !== undefined)  {
+          handleClick(data);
+        }
+      }}
+    >
+      <DraggableRowWrapper
+        index={index || 0}
```
Comment (juliensnz) [inline src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/shared/datagrids/Row.tsx:32]: same here, maybe we could put the default value in the spread instead?

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/shared/layouts/tables/TableHeadRow.tsx [hunk 17]
```diff
@@ -0,0 +1,14 @@
+import styled from 'styled-components';
+
+const TableHeadRow = styled.tr`
+  cursor: auto;
+  height: calc(44px + 15px);
```
Comment (juliensnz) [inline src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/shared/layouts/tables/TableHeadRow.tsx:5]: Why do we do a calc here? Maybe you could add a comment to give some context?

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/hooks/attribute-groups/useGetAttributeGroupLabel.ts [hunk 18]
```diff
@@ -0,0 +1,18 @@
+import {useCallback} from 'react';
+import {useUserContext} from '@akeneo-pim-community/legacy-bridge';
+import {AttributeGroup} from '../../models';
+
+// @ts-ignore
```
Comment (juliensnz) [inline src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/hooks/attribute-groups/useGetAttributeGroupLabel.ts:5]: Why do we nned this ts ignore?

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/infrastructure/fetchers/attributeGroupsFetcher.ts [hunk 19]
```diff
@@ -0,0 +1,14 @@
+import {AttributeGroupCollection} from '../../models';
+
+const FetcherRegistry = require('pim/fetcher-registry');
+
+const fetchAllAttributeGroups = async (): Promise<AttributeGroupCollection> => {
+  try {
+    return FetcherRegistry.getFetcher('attribute-group').fetchAll();
+  } catch (error) {
+    console.error(error);
+    return Promise.resolve({});
```
Comment (juliensnz) [inline src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/infrastructure/fetchers/attributeGroupsFetcher.ts:10]: It should reject no?

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/infrastructure/savers/attributeGroupsSaver.ts [hunk 20]
```diff
@@ -0,0 +1,30 @@
...
+      ],
+      body: JSON.stringify(sortOrder),
+    }).then(response => {
+      FetcherRegistry.getFetcher('attribute-group').clear();
+
+      return response.json();
+    });
+  } catch (error) {
+    console.error(error);
+    return Promise.resolve({});
```
Comment (juliensnz) [inline src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/infrastructure/savers/attributeGroupsSaver.ts:26]: same here, it should reject no?

**Review by juliensnz — COMMENTED** [review]
Really nice improvements :) Added small comments

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/shared/layouts/tables/TableHeadRow.tsx (hunk 17 above)
Comment (olivierpontier) [reply src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/components/shared/layouts/tables/TableHeadRow.tsx:5]: A brainless copy/past, sorry about that

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/hooks/attribute-groups/useGetAttributeGroupLabel.ts (hunk 18 above)
Comment (olivierpontier) [reply src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/hooks/attribute-groups/useGetAttributeGroupLabel.ts]: PhpStorm can not resolve the path of the library `pimui/js/i18n`. It is not mandatory at all.

File: src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/infrastructure/fetchers/attributeGroupsFetcher.ts (hunk 19 above)
Comment (olivierpontier) [reply src/Akeneo/Pim/Structure/Bundle/Resources/workspaces/settings-ui/src/infrastructure/fetchers/attributeGroupsFetcher.ts:10]: I log the error but I want the UI works with an empty result. I don't know if it is the good way to do it.

**Pull request merged**